src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java

```diff
@@ -638,131 +609,125 @@ static PartiallyDeterminedHaplotype createNewPDHaplotypeFromEvents(final Haploty
             //TODO we know there is an optimization involving minimizing the number of checks necessary here by iterating
             //TODO using the bitmask values themselves for the loop
-            if (!forbiddenCombinations.isEmpty()) {
+            if (!mutexes.isEmpty()) {
                 for (final SmallBitSet subset = new SmallBitSet().increment(); !subset.hasElementGreaterThan(eventsInOrder.size()); subset.increment()) {
-                    if (forbiddenCombinations.stream().anyMatch(subset::contains)) {
-                        allowedEvents.set(subset.index(), false);
+                    if (mutexes.stream().anyMatch(subset::contains)) {
+                        allowedSubsets.set(subset.index(), false);
                     }
                 }
             }
         }
 
         /**
-         * This method handles the logic involved in getting all of the allowed subsets of alleles for this event group.
+         * For a given set of determined events, find all subsets of the event group such that each subset
+         *
+         * 1) contains all determined events that overlap this event group
+         * 2) contains no two or three mutually exclusive elements i.e. is an allowed subset as computed in this object's constructor
+         * 3) is not a subset of any larger subset that satisfies 1) and 2)
+         *
+         * Note that the subsets need not be disjoint and need not cover all events in this EventGroup.
+         *
+         * Because these subsets contain no mutexes, one can sensibly make partially-determined haplotypes out of them,
+         * with each subset entailing a unique set of undetermined alleles.  Property 3) above ensures that we make as
+         * few partially-determined haplotypes as possible.
          *
-         * @param disallowSubsets
          * @return
          */
-        public List<Set<Event>> setsForBranching(final List<Event> locusEvents, final Set<Event> determinedEvents, final boolean disallowSubsets) {
-            final SmallBitSet locusOverlapSet = overlapSet(locusEvents);
-            final SmallBitSet determinedOverlapSet = overlapSet(determinedEvents);
+        public List<Set<Event>> eventSetsForPDHaplotypes(final Set<Event> determinedEvents) {
+            if (eventsInOrder.size() == 1) {
+                return Collections.singletonList(Set.of(eventsInOrder.get(0))); // if only one event, there are no mutexes
+            }
+            final SmallBitSet determinedSubset = overlapSet(determinedEvents);
 
-            // Special case (if we are determining bases outside of this mutex cluster we can reuse the work from previous iterations)
-            if (locusOverlapSet.isEmpty() && cachedEventSets != null) {
+            // We use a cache for the recurring case where the determined events do not belong to this event group
+            if (determinedSubset.isEmpty() && cachedEventSets != null) {
```

Comment:
i'm not sure this logic is correct anymore? Determined subset will usually never be empty here because its coming from the list of alleles and you haven't actually checked that it is or isn't present in this event group.

Reply:
i think its safer to put this subset checking code inside of this method rather than outside

Reply:
I improved this part.